src/hooks/useGlobalIntersectionObserverWithDelay.ts

```diff
@@ -0,0 +1,125 @@
+   * is visible for delayForMS number of seconds.
+   */
+  const timeouts = useRef<Record<any, number>>({});
+  const updateEntry = useCallback(
+    (entries: IntersectionObserverEntry[]): void => {
+      entries.forEach((visibleEntry) => {
+        let elementIntersectionDebuggerNode: HTMLElement;
+        if (isDev) {
+          const nodes = visibleEntry.target.getElementsByClassName('debugger');
+          // if at least one debugger node was found (should always be there but just to be extra sure)
+          if (nodes.length) {
+            elementIntersectionDebuggerNode = nodes[0] as HTMLElement;
+          }
+        }
+        const targetNode = visibleEntry.target as HTMLElement;
+        const elementId = targetNode.dataset[elementIdKey];
+        // if the element content container is interacting
+        if (visibleEntry.isIntersecting) {
+          setIntersectionDebuggerNodeText(
+            `Element is fully visible, triggering callback in ${delayForMS / 1000} seconds...`,
+            elementIntersectionDebuggerNode,
+          );
+          // Call the API after delayForMS
+          timeouts.current = {
+            ...timeouts.current,
+            [elementId]: window.setTimeout(() => {
+              onElementVisible(visibleEntry.target);
+              setIntersectionDebuggerNodeText(
+                `Element callback triggered`,
+                elementIntersectionDebuggerNode,
+              );
+            }, delayForMS),
+          };
+        } else {
+          // if it's no longer intersecting, clear the timeout
+          clearTimeout(timeouts.current[elementId]);
+          /**
+           * Delete the timeout function's ID from the global array since it's no longer
+           * needed and next time if the same element intersects, a new ID will be generated
+           * for that element.
+           */
+          const tempTimeouts = { ...timeouts.current };
+          delete tempTimeouts[elementId];
+          timeouts.current = tempTimeouts;
+          setIntersectionDebuggerNodeText(
+            `Element is not fully visible, won't trigger callback`,
+            elementIntersectionDebuggerNode,
+          );
+        }
+      });
+    },
+    [delayForMS, elementIdKey, onElementVisible],
+  );
+
+  /**
+   * We need to assign to window before render is done since we want to the observer to be available by the time rendering is done so that the elements to be observed can use it.
+   * {@see https://kentcdodds.com/blog/useeffect-vs-uselayouteffect}
+   */
+  useBrowserLayoutEffect(() => {
+    const hasIOSupport = !!window.IntersectionObserver;
+    if (!hasIOSupport) return undefined;
+
+    // no need to create a new observer if it already exists.
+    if (!window[observerId]) {
+      window[observerId] = new IntersectionObserver(updateEntry, {
+        threshold,
+        root,
+        rootMargin,
+      });
+    }
+    return () => {
```

Comment:
We don't clear timeouts here, causing requests to be sent even if you closed the reflection modal. Loop over all `timeouts`, clear them, and empty the timeouts `Map` after.

Reply:
Yes you are right my bad let me do that.

Reply:
I just double-checked. Closing the reflections modal will not trigger the clean-up function since the component `ReflectionBodyContainer` will still be rendered even after you close the modal and closing it will cause all the timeout functions to be cleared anyways since none of the elements be observed will be intersecting. I will add a manual clear timeout thought just to be extra sure.